Update syntax with examples, callables and wording

maxfischer2781 opened the pull request:
This pull request includes:

* categories for constructs (e.g. Containers, Callable, ...),
* example values to most types, e.g. ``foo = (1, 'two', 3.0)`` for ``foo: "(int, str, float)"``.
* new ``(a: int, *float, b: bool, **str)``Callable syntax for variadic and named arguments.

Since the Callable syntax is a proposal, it is currently in ``.. hint::`` blocks.

maxfischer2781 committed:
Update syntax with examples, callables and wording

Add example values to most type examples, e.g. ``foo = (1, 'two', 3.0)`` for ``foo: "(int, str, float)"``. Also adds Callable syntax for variadic and named arguments.

a-recknagel commented on docs/syntax.rst:
```diff
@@ -55,6 +55,11 @@ similar. Given the examples, they would change in the following ways:
 ``typing``-Types Equivalent Mappings
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+Sepcial Forms
```
typo

a-recknagel replied:
also, the sub-sub sections help a lot visually, it was a good idea to add them.

maxfischer2781 replied:
Thanks. We should probably reshuffle the sections in the future - didn't want to the diff to become meaningless.

a-recknagel committed:
The CI/CD also needs to run for contributions.

That was the whole point, how could I forget to add that??

a-recknagel commented on docs/syntax.rst:
```diff
@@ -147,36 +166,83 @@ value types. Use literal notation in ``stenotype`` mode.
+  foo: "(str, int) -> int"
 
-----
+.. note::
```
the note block doesn't render correctly in the github rst view, so I don't know if it looks good. Could you rebase on master or merge it back in? I changed the CI to run on PRs so that I can take a look at reference artifacts.

a-recknagel commented on docs/syntax.rst:
```diff
@@ -236,44 +305,63 @@ ForwardRef_
 
+  @contextmanager
+  def foo(a: A, b: B):
+      yield r
```
this should return, right?

maxfischer2781 replied:
It's a [``contextlib.contextmanager``](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) which does need the ``yield`` -- do you think that is too obscure? I don't know if there are any well-known context managers in the standard library.

a-recknagel commented on docs/syntax.rst:
```diff
@@ -236,44 +305,63 @@ ForwardRef_
   # stenotype
-  "(A, B) -> await *R"
+  foo: "(A, B) -> await *R"
+  bar: "await *R"
```
I think I prefer `iter R` instead of `*R`. The problem with `for R` is that it sound a little silly in conjunction with, for example, context managers . i.e., rather `await iter R` than `await for R`.

maxfischer2781 replied:
Agreed. Will add that in the PR.

a-recknagel commented on docs/syntax.rst:
```diff
@@ -97,15 +102,20 @@ is what a few other languages allow as well.
 
 .. code-block:: python
+  
+  # value
```
with `value` blocks as well these sections start looking a little busy. I was thinking of including [sphinxcontrib-contentui](https://sphinxcontrib-contentui.readthedocs.io/en/latest/tabs.html) to use its tab-boxes anyway for this page, now that most of them have three tabs I think it's necessary. It's not a blocker for this PR, but you can add it now already if you like.

maxfischer2781 replied:
I like the idea of tab boxes - that would also allow a proper grammar alongside examples. Let's make that a separate pull request, though.

a-recknagel reviewed (commented):
Looks good, all the changes make sense.

maxfischer2781 commented on docs/syntax.rst:
```diff
@@ -55,6 +55,11 @@ similar. Given the examples, they would change in the following ways:
 ``typing``-Types Equivalent Mappings
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+Sepcial Forms
```
```suggestion
Special Forms
```

maxfischer2781 commented on docs/syntax.rst:
```diff
@@ -188,7 +254,7 @@ when using ``stenoype``.
   foo: "*int"
```
```suggestion
  foo: "iter int"
```

maxfischer2781 commented on docs/syntax.rst:
```diff
@@ -188,7 +254,7 @@ when using ``stenoype``.
   foo: "*int"
   bar: "() -> *bool"
```
```suggestion
  bar: "() -> iter bool"
```

maxfischer2781 commented on docs/syntax.rst:
```diff
@@ -236,44 +305,63 @@ ForwardRef_
 
   # stenotype
-  "(A, B) -> await *R"
+  foo: "(A, B) -> await *R"
```
```suggestion
  foo: "(A, B) -> await iter R"
```

maxfischer2781 commented on docs/syntax.rst:
```diff
@@ -236,44 +305,63 @@ ForwardRef_
   # stenotype
-  "(A, B) -> await *R"
+  foo: "(A, B) -> await *R"
+  bar: "await *R"
```
```suggestion
  bar: "await iter R"
```

maxfischer2781 committed:
Apply suggestions from code review

maxfischer2781 commented:
I've fixed the typo of the first section and renamed ``*R`` to ``iter R`` on all occurrences I found. 

👍 on adding content tabs, but I recommend doing that in a separate PR.
